move software testing lesson (#1)
esciencecenter-digital-skills/good-practices-lesson: episodes/testing.md

---
title: Testing
teaching: 35
exercises: 25
---

:::::::::::::::::::::::::::::::::::::: questions 

> - Why test?

Reviewer: ```suggestion
- Why should I write automated tests for my code?
- How do I write a good unit test?
```

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

> - Understand the place of testing in a scientific workflow.
> - Understand that testing has many forms.

Reviewer: ```suggestion
- Use `pytest` to write and run unit tests
- Understand that testing has many forms.
```

 
 
def read_data(file_name):
    return pd.read_csv(file_name)
 
 
def get_random_number(number):
    return random.random()
```

And not so intuitive:

```python
def hello(name):
    print("Hello", name)
 
 
nums = [1, 2]
 
def append(a_list, item):
    a_list += [item]
    return a_list
 
print(nums)            # [1, 2]
print(append(nums, 3)) # [1, 2, 3]
print(nums)            # [1, 2, 3] 😬
```

Some side effects can be indeed necessary or hard to spot.

Use pure functions when possible 👌

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: challenge 

## Test-Driven Development: FizzBuzz Function (15 min)

The function `fizz_buzz()` takes an integer argument and returns it, BUT:

- Fails on zero or negative numbers.
- Instead returns "Fizz" on multiples of 3.
- Instead returns "Buzz" on multiples of 5.
- Instead returns "FizzBuzz" on multiples of 3 and 5.

Create an empty function `fizz_buzz()` and go through the conditions listed above, one by one:

1. Write a test for the condition.
2. Edit the `fizz_buzz()` function until the test passes.

Then discuss together the different solutions.

:::::::::::::::::::::::: solution 

## Solution
 

> Add here the preferred solution. We don't have a ready one. 

Reviewer: ```suggestion
We are still working on example code for a solution. But as long as you added tests for all the different conditions and your final `fizz_buzz()` function passes all of them your solution is correct.
```

Did you create an issue for this?
